tests/lang/s02_core/s01_procedures/t02_argument_passing.nim

```diff
@@ -0,0 +1,318 @@
+    type
+      Object = object
+        field: int
+
+    proc useOther(obj: Object, arg: int = obj.field): (int, int) = (obj.field, arg)
+
+    doAssert useOther(Object(field: 1)) == (1, 1)
+    doAssert useOther(Object(field: 1), 2) == (1, 2)
+
+
+  block call_procedure:
+    proc getValue(): int = 12
+
+    proc useOther(arg: int = getValue()): int = arg
+
+    doAssert useOther(getValue()) == useOther()
+
+    doAssert useOther() == 12
+    doAssert useOther(1) == 1
+
+  block evaluate_block_expression:
+    proc useOther(
+      arg: int = 12,
+      other: int =
+        block:
+          arg
+    ): (int, int) = (arg, other)
+
+    doAssert useOther() == (12, 12)
+    doAssert useOther(2) == (2, 2)
+    doAssert useOther(1, 2) == (1, 2)
+
+
+
+
+block varargs_arguments_no_converter:
+  block single_varargs:
+    proc impl(vas: varargs[int]): seq[int] = @vas
+
+
+    doAssert impl() == newSeq[int]()
+    doAssert impl(1) == @[1]
+    doAssert impl(1, 2, 4, 5, 6, 7) == @[1, 2, 4, 5, 6, 7]
+
+  # Multiple varargs is most likely a feature that is not used widely
+  # (if at all), but this is stil supported, so should probably be tested.
+  block multiple_varargs:
+    proc impl(va1: varargs[int], va2: varargs[bool]): (seq[int], seq[bool]) =
+      return (@va1,  @va2)
+
+    doAssert impl(1) == (@[1], newSeq[bool]())
+
+    if false:
+      # Right nor results in `(@[1, 2, 0], @[true, false, false])`
+      doAssert impl(1, 2, false) == (@[1, 2], @[false])
+
+    ## It is not possible to entirely omit leading varargs arguments, but it
+    ## is possible to explicitly pas values to the other varargs if present
+    doAssert impl(va2 = true) == (newSeq[int](), @[true])
+
+
+  block command_syntax:
+    proc vaProc(args: varargs[int]): seq[int] = @args
+
+    block:
+      ## Command syntax cannot be used in the `let val = cmd 1,2,3,4,4`
+      ##
+      ## REVIEW is it a bug?
+      let val =
+        block:
+          vaProc 1, 2, 3
+
+      doAssert val == @[1,2,3]
+
+    block:
+      ## When used without any arguments in the `let val = vaProc`
+      ## expressions, procedure itself gets assigned to the variable.
+      let val = vaProc
+      doAssert val is proc
+
+
+  block regular_syntax:
+    proc vaProc(args: varargs[int]): seq[int] = @args
+
+    doAssert vaProc(1, 2, 3) == @[1, 2, 3]
+    doAssert vaProc(1, 2) == @[1, 2]
+    doAssert vaProc() == newSeq[int]()
+
+  block block_syntax:
+    proc vaProc(args: varargs[int]): seq[int] = @args
+
+    doAssert():
+      `==`():
+        @[12]
+
+      do:
+        vaProc():
+          12
+
+  block block_syntax_do:
+    proc vaProc(args: varargs[int]): seq[int] = @args
+
+    doAssert():
+      `==`():
+        @[12, 13]
+
+      do:
+        vaProc():
+          12
+
+        do:
+          13
+
+
+block varargs_arguments_with_converter:
+  proc convertToInt(arg: float): int = int(arg)
+
+  proc vaWithConverter(args: varargs[int, convertToInt]): seq[int] = @args
```

Comment:
Todo, now for later, revisit the interaction here because this can seriously mess up overload resolution.